cmd/restic/cmd_rebuild_index.go

```diff
@@ -44,58 +52,109 @@ func runRebuildIndex(gopts GlobalOptions) error {
 
-	var packs uint64
-	err := repo.List(ctx, restic.PackFile, func(restic.ID, int64) error {
-		packs++
+	var obsolete restic.IDs
+	err := repo.List(ctx, restic.IndexFile, func(id restic.ID, size int64) error {
+		obsolete = append(obsolete, id)
 		return nil
 	})
 	if err != nil {
 		return err
 	}
 
-	bar := newProgressMax(!globalOptions.Quiet, packs-uint64(len(ignorePacks)), "packs")
-	idx, invalidFiles, err := index.New(ctx, repo, ignorePacks, bar)
-	if err != nil {
-		return err
-	}
+	packs := restic.NewIDSet()
+	packsWronglyIndexed := restic.NewIDSet()
+	totalBlobs := uint(0)
+
+	if opts.ReadAllPacks {
+		Verbosef("finding pack files in repo\n")
+		err = repo.List(ctx, restic.PackFile, func(id restic.ID, size int64) error {
+			packs.Insert(id)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	} else {
+		Verbosef("load indexes\n")
+		err = repo.LoadIndex(gopts.ctx)
+		if err != nil {
+			return err
+		}
+
+		type PackInfo struct {
+			size  uint64
+			blobs uint
+		}
+		packInfo := make(map[restic.ID]PackInfo)
+
+		Verbosef("get pack files to read\n")
+		// iterate over all blobs in index
+		for blob := range repo.Index().Each(ctx) {
+			pi, ok := packInfo[blob.PackID]
+			if !ok {
+				pi.size = pack.HeaderSize
+			}
+			pi.size += uint64(pack.PackedSizeOfBlob(blob.Length))
+			pi.blobs++
+			// update packSizeFromIndex
+			packInfo[blob.PackID] = pi
+		}
 
-	if globalOptions.verbosity >= 2 {
-		for _, id := range invalidFiles {
-			Printf("skipped incomplete pack file: %v\n", id)
+		err := repo.List(ctx, restic.PackFile, func(id restic.ID, packSize int64) error {
+			pi, ok := packInfo[id]
+			if !ok || pi.size != uint64(packSize) {
+				// Pack was not referenced in index or size does not match
+				packs.Insert(id)
+			} else {
+				totalBlobs += pi.blobs
+			}
+			delete(packInfo, id)
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		// process pack files that are referenced in the index but do not exist
+		for id := range packInfo {
+			packsWronglyIndexed.Insert(id)
 		}
 	}
 
-	Verbosef("finding old index files\n")
+	if len(packs) > 0 {
+		Verbosef("reading pack files...\n")
+		blobsBefore := countBlobs(repo)
+		bar := newProgressMax(!globalOptions.Quiet, uint64(len(packs)), "packs")
+		err = repo.LoadIndexFromPacks(ctx, packs, obsolete, bar)
+		if err != nil {
+			return err
+		}
+		blobsAfter := countBlobs(repo)
+		totalBlobs += blobsAfter - blobsBefore
 
-	var supersedes restic.IDs
-	err = repo.List(ctx, restic.IndexFile, func(id restic.ID, size int64) error {
-		supersedes = append(supersedes, id)
-		return nil
-	})
-	if err != nil {
-		return err
-	}
+		//	if globalOptions.verbosity >= 2 {
+		//		for _, id := range invalidFiles {
+		//			Printf("skipped incomplete pack file: %v\n", id)
+		//		}
+		//	}
 
-	ids, err := idx.Save(ctx, repo, supersedes)
-	if err != nil {
-		return errors.Fatalf("unable to save index, last error was: %v", err)
+	} else {
+		Verbosef("no need to read any pack file.\n")
```

Comment:
Please remove the "." . The other output lines also don't have one.

Reply:
done.